### PR TITLE
feat(server): add ability to delete project, fixes #4044

### LINKED ISSUE #4044: 🚀 Feature: Allow users to delete a project
### Feature description

- Add a new delete button on the Project Configuration item

image

- When clicked, show the user a warning with a confirmation button "This action cannot be undone. This will permanently delete the project and all its resources. Are you sure you want to continue?"

image

- On the server side, use soft delete (there is already a "deletedAt" field). Copy the behavior from resource.service.ts

### Use case

The user wants to delete a project

### Are you willing to submit PR?

_No response_

### PR DESCRIPTION
**Pull request by MichaelSolati**
Close: #4044

## PR Details

- Add a mutation to the project resolver for deleting a project
- Add a method to project service to delete a project
  - First archives all resources belonging to a project
  - Update billing service based on archived services
  - Next "deletes" project by renaming it and setting `deletedAt` field
- Add `archiveProjectResources` method to the resource service
  - Gets all resources belonging to a project and then checks if they're empty
  - Batches mutations of resources and marks them as "archived"
  - Takes batch and runs it as a transaction (trying to limit Prisma calls)
- Update UI to call `deleteProject` mutation

## PR Checklist

- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

**Comment by MichaelSolati** [comment]
Sorry to add so many of y'all for the review, I'm just unsure who's responsible for these files/services/etc

File: packages/amplication-server/src/core/resource/resource.service.ts [hunk 1]
```diff
@@ -426,6 +426,67 @@ export class ResourceService {
...
+      })
+    );
+    const deletedResources = await this.prisma.$transaction(deleteResources);
+
+    if (this.isBillingEnabled) {
+      await this.billingService.reportUsage(
+        workspace.id,
+        BillingFeature.Services,
+        deletedServiceCount
+      );
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resource.service.ts:475]: we may need to keep this code anyway even if we not actually deleting the resources, so maybe you could move it to the "deleteProject" function

File: packages/amplication-server/src/core/resource/resource.service.ts [hunk 2]
```diff
@@ -426,6 +426,67 @@ export class ResourceService {
...
+    }
+
+    await Promise.all(
+      resources.reduce((all: Promise<Resource>[], resource) => {
+        if (resource.gitRepositoryOverride) {
+          all.push(this.deleteResourceGitRepository(resource));
+        }
+        return all;
+      }, [])
+    );
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resource.service.ts:485]: we should keep this data also as it is (for future undo)

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
well done @MichaelSolati 
It is great to see how you followed all the best practices and navigated your way on the code. 

Please see my comments regarding leaving the resources in place

**Comment by MichaelSolati** [comment]
@yuval-hazaz just to secure a solution to the resources. Should we block resources from being queried/modified by making checking the status of the parent project, or should we have a status field on a resource as to what the parent's status is?

I'm inclined to prefer adding a status field to the resource, we could just call it `parentsStatus` and allow it to be `null` or `'deleted'`. Down the road we can add new statuses such as `archived'`.

Let me know what you think.

**Comment by yuval-hazaz** [comment]
> I'm inclined to prefer adding a status field to the resource, we could just call it `parentsStatus` and allow it to be `null` or `'deleted'`. Down the road we can add new statuses such as `archived'`.

I am OK with adding a new field, but in this case, I prefer to keep it simple and use a boolean with a specific definition - it will hold a boolean value whether the record is effectively active or not. It should hold "false" whether the actual record is deleted or any of its parents is deleted. 

When fetching the data, we should only check for that value. 

we can call the field "EffectiveActive"

**Comment by abrl91** [comment]
@MichaelSolati, any progress?

**Comment by MichaelSolati** [comment]
Working on this today. Should be done in the next day or so.

**Comment by MichaelSolati** [comment]
I am finished making the requested changes. I don't think the name `effectiveActive` is that clear as to what it represents; I'm wondering if we can rename it to `archived`?

**Comment by yuval-hazaz** [comment]
> I am finished making the requested changes. I don't think the name `effectiveActive` is that clear as to what it represents; I'm wondering if we can rename it to `archived`?

I think "archived" can work

**Comment by MichaelSolati** [comment]
@yuval-hazaz ready for review

**Review by dulajdilshan — COMMENTED** [review]
LGTM

**Comment by MichaelSolati** [comment]
@abrl91 @yuval-hazaz @tupe12334 

Can someone review this please.

**Comment by abrl91** [comment]
It looks good, I'm just wondering what happens when all projects are deleted, I mean, what will happen to the UI?

image

File: packages/amplication-server/src/core/project/project.service.ts [hunk 3]
```diff
@@ -72,6 +75,39 @@ export class ProjectService {
...
+      throw new Error(INVALID_PROJECT_ID);
+    }
+
+    const archivedResources =
+      await this.resourceService.archiveProjectResources(project.id);
+    const archivedServiceCount = archivedResources.filter(
+      (r) => r.resourceType === EnumResourceType.Service
+    ).length;
+
+    if (this.billingService.isBillingEnabled) {
```
[the remark this answers is not in the document]
Comment (MichaelSolati) [reply packages/amplication-server/src/core/project/project.service.ts:94]: Beautiful, I'll remove this check then.

File: packages/amplication-server/migration-scripts/install-auth-core-plugin.ts [hunk 4]
```diff
@@ -71,6 +71,7 @@ async function main() {
     where: {
       resourceType: EnumResourceType.Service,
       deletedAt: null,
+      archived: { not: true },
```
Comment (abrl91) [inline packages/amplication-server/migration-scripts/install-auth-core-plugin.ts:74]: We already ran this migration script on the DB, so I don't this is relevant.

**Comment by abrl91** [comment]
@MichaelSolati, generally, it looks good to me, but I wonder if we want to merge it before fixing the UI issue in case a user deletes all the projects.

As for your question, I am trying to understand. Can you create an issue with more details/examples?

**Review by abrl91 — APPROVED** [review]

**Review by yuval-hazaz — APPROVED** [review]

**Pull request merged**